storage/src/mmr/mem.rs

```diff
@@ -0,0 +1,854 @@
+/// with root hash `root_hash`.
+pub fn verify_proof<const N: usize, H: Hasher<N>>(
+    proof: &Proof<N>,
+    element: &Hash<N>,
+    element_pos: u64,
+    root_hash: &Hash<N>,
+    hasher: &mut H,
+) -> bool {
+    let mut peak_hashes = Vec::new();
+    let mut proof_hashes = proof.hashes.iter();
+    let mut missing_peak_height = u32::MAX;
+
+    for (peak_pos, height) in PeakIterator::new(proof.sz) {
+        if missing_peak_height == u32::MAX && peak_pos >= element_pos {
+            // Found the peak of the tree whose hash we must reconstruct.
+            missing_peak_height = height;
+            let mut hashes = proof.hashes.iter().rev();
+            match peak_hash_from_element(
+                peak_pos,
+                height,
+                element,
+                element_pos,
+                &mut hashes,
+                hasher,
+            ) {
+                Ok(peak_hash) => peak_hashes.push(peak_hash),
+                Err(_) => return false, // missing proof hashes
+            }
+        } else if let Some(peak_hash) = proof_hashes.next() {
+            peak_hashes.push(*peak_hash);
+        } else {
+            return false; // invalid number of proof hashes
+        }
+    }
+
+    // To prevent proof malleability, verify that every proof hash must have been used as input.
+    if proof.hashes.len() != peak_hashes.len() - 1 + (missing_peak_height as usize) {
+        return false;
+    }
+    *root_hash == hasher.root_hash(proof.sz, peak_hashes.iter())
+}
+
+/// Computes the peak hash of the perfect tree with peak `peak_pos` from one of its elements and its
+/// sibling hashes along the path from leaf to peak, returning the result unless some hashes were
+/// missing.
+fn peak_hash_from_element<'a, const N: usize, H: Hasher<N>>(
+    peak_pos: u64,
+    height: u32,
+    element: &Hash<N>,
+    element_pos: u64,
+    sibling_hashes: &mut impl Iterator<Item = &'a Hash<N>>,
+    hasher: &mut H,
+) -> Result<Hash<N>, ()> {
+    let mut hash = hasher.leaf_hash(element_pos, element);
+    let path_vec: Vec<(u64, u64)> =
+        PathIterator::new(element_pos, peak_pos, height).collect::<Vec<_>>();
+    let mut two_h = 2;
+
+    for (parent_pos, pos) in path_vec.into_iter().rev() {
+        let sibling_hash = sibling_hashes.next().ok_or(())?;
+        hash = if parent_pos == pos + 1 {
+            hasher.node_hash(pos + 1, &hash, sibling_hash)
+        } else {
+            hasher.node_hash(pos + two_h, sibling_hash, &hash)
+        };
+        two_h <<= 1;
+    }
+    Ok(hash)
+}
+
+/// Return true if `proof` proves that the `elements` appear consecutively between positions
+/// `start_element_pos` through `end_element_pos` (inclusive) within the MMR with root hash
+/// `root_hash`.
+pub fn verify_range_proof<const N: usize, H: Hasher<N>>(
+    proof: &Proof<N>,
+    elements: &[Hash<N>],
+    start_element_pos: u64,
+    end_element_pos: u64,
```

Comment:
I think we are safe to infer this from the length of elements?

Reply:
The elements are just the leaves, the element positions account for nodes between them. So the inference would involve extra computation, which seems unnecessary given the positions are the primary identifier for elements?

Reply:
Ahh duh. Disregard.

Reply:
moved the MMR position example from the test cases to the header and explicitly noted that elements may not have contiguous positions due to the postfix node ordering.